contrib/upgradability-patterns.asciidoc

```diff
@@ -0,0 +1,367 @@
+////
+Sources: https://github.com/SylTi/zeppelin-solidity/commit/ca947d89ada67b09c3590b754ea7a83e624e93d2
+https://blog.zeppelin.solutions/proxy-libraries-in-solidity-79fbe4b970fd
+License: MIT
+Added By: @SylTi
+////
+
+[[upgradable_smart_contracts_chapter]]
+== Upgradable Smart Contracts
+
+[[why]]
+=== Advantages of using Upgradable smart contracts
+
+Ethereum smart contracts are immutable and as such can't be modified once they are deployed. If you have a bug, or a missing feature once the smart contract is in production, this can become a very serious problem. Think of the DAO, and Parity hacks to name a few.
+
+Deploying a new version of the contract and importing it's data to a new address, in the cases where it's even a possible solution to the problem, is very costly if not outright impossible due to the gasPrice and gasLimit limiting the amount of data than a contract can copy during its creation.
+
+Fortunately you can use programming patterns proactively inside your smart contracts to remove those limitations from occurring later on.
+
+Those patterns can be regrouped into 2 family: Eternal Storage and Proxy libraries.
+
+
+=== Drawbacks
+Have we have seen, having the possibility of upgrading smart contract is a useful property but it is also a double-edged sword. It negates the promise of "Code is Law" by reintroducing a central point of failure and new attack vectors.
```

Comment:
Please, some native English contributor could validate the **"Have we have seen"** phrase?
Me, as a non native Engliish contributor, don't finish seeing it well constructed grammatically.

Reply:
Awesome sentence --> **"It negates the promise of "Code is Law" by reintroducing a central point of failure and new attack vectors"**.